### PR TITLE
2235 Added Webhook logs section and Webhook UI/UX enhancements

### PR DESCRIPTION
**Pull request by albertisfu**
The webhook logs section is added here and also #2367 UI/UX enhancements were applied.

Now the Webhooks page is sub-divided into 3 tabs: 
- Webhooks (Where user can manage their endpoints)
- Logs (Production logs)
- Test Logs (Testing Logs)

**Logs section:**
- Logs and Test Logs sections include a table with webhook events ordered by `Date created` (latest first), including pagination (20 elements by page).
- Filters by `Event Type` and `Event Status` were implemented (also paginated).

Screen Shot 2022-11-18 at 10 53 54

Screen Shot 2022-11-18 at 10 54 27

Screen Shot 2022-11-18 at 10 54 35
 
The webhook event log detail includes information about the Webhook endpoint and Webhook Event:
- Endpoint URL
- Endpoint Status
- Event Type
- Event ID (idempotency-Key)
- Date created
- Status code (If any)
- Event Status
- Try counter
- Next retry date
- Content
- Response (If any)
- Error message (If any)

Screen Shot 2022-11-18 at 10 55 43

Webhook event log detail for testing events:
Screen Shot 2022-11-18 at 10 55 59

A couple of mobile screenshots:
IMG_9942

IMG_9943

#2367 Enhancements:

- Updated `<title>` for API and Webhook pages.
- Updated `<title>` for Alerts pages.

- Fixed Edit Webhook modal title.
- Fixed order in Webhook endpoints buttons.
- Added the Failure count to Webhook endpoints

Screen Shot 2022-11-18 at 10 54 45

Test modal:
- Fixed textareas indentation.
- The button for cURL moved on the same line as the sending the event button.
- Fixed text for: Send Webhook Test Event
- When a user clicks the `Send Webhook Test Event` button the button for Tests logs is displayed and linked to the Test Logs section.
- Cancel button text changed to Close
- Fixed URL in the cURL command wrapped in single quotes.

Screen Shot 2022-11-18 at 11 23 57

Let me know what you think!

### CODE REVIEW

File: cl/users/templates/includes/webhook-event-detail.html
Comment (mlissner) [inline cl/users/templates/includes/webhook-event-detail.html:70]: Surprised to see this here. Should be abstracted out.

File: cl/users/templates/includes/webhooks_htmx/webhooks-pagination.html
Comment (mlissner) [inline cl/users/templates/includes/webhooks_htmx/webhooks-pagination.html:8]: Can this pagination file be merged with the normal include, so it's not so much duplicated code? I guess you tried that, but it wasn't worth it?

File: cl/users/templates/profile/webhooks_logs.html
Comment (mlissner) [inline cl/users/templates/profile/webhooks_logs.html:67]: DRY, please.

File: cl/users/views.py
Comment (mlissner) [inline cl/users/views.py:778]: Let's merge these together, please.

File: cl/users/views.py
Comment (mlissner) [inline cl/users/views.py:793]: This is cool. We're getting some nice benefits from our design choices.

**Review by mlissner — CHANGES_REQUESTED** [review]
Whew, big one. A few UI suggestions first:

1. Instead of saying `None` in the logs page when we lack a Status Code or a Next Event, could we just put a dash? I think it'd de-clutter things a bit.

2. The column header for `Date Created` and the field for "Date Created" in the "Event Detail" page — Would that make more sense as "Event Date"? 

3. On the "Event Detail" page, instead of Event ID should it say "Idempotency-key"? 

4. On the "Event Log Pages" why not make the whole row clickable instead of having a "Details" button?

5. Next to our Status Code fields, could we put the Python ENUM name: https://docs.python.org/3/library/http.html#http-status-codes? Helpful?

6. One of your screenshots of the event details shows "Delivered Successfully" with a "Try Counter" of 7. I guess that means it took seven tries to deliver it? Two thoughts. First, where are the other six responses that failed? They should be in the logs somewhere, right? Second, maybe this field should be called "Failure Count" instead, and it'd be six instead of seven, I guess?

7. I think the header "Webhook event detail:" should be capitalized, pluralized. and the colon removed: "Webhook Event Details"

8. The failure count that you added to the Webhooks page, does that reset to zero when the webhook is working or does it only go up, up up?

I ran into this issue when clicking the logs tab without any webhooks in place:

image

The error was b/c I needed to migrate my DB (oops), but the English should be tweaked to:

> We ran into an error processing your request and will be looking into it. Please try again later.

9. I'm sorry, I think the test modal should be better, and I think the new "Sent it, view in logs" button, although useful and cool, is a bit confusing. Could you redo this just a bit so that:

     - The top is the same.
     - There are two tabs. The first says "As JSON", the other is "As cURL". 
     - In the first panel, it shows the textarea for the JSON, like now, and at the bottom of the panel the button is there to send the event. A second button is there that says, "View Test Logs", and just takes the user to the list of test logs.
     - The second panel has the textarea for the cURL command, and has a button at the bottom that says "Copy cURL Command".
     - When the user clicks the "Send event" button, it can do a spinner icon that becomes a checkbox when the event is successfully sent, instead of showing the new green button.

    I think this would be clearer and more compact too. 
-----

As far as code goes, I looked at everything but the JS. Once the above is done, I'll do a second review and, well, you know, I'll probably have more comments. 

All in all, this looks great though. Really nice system coming together.

File: cl/assets/static-global/css/override.css
Comment (albertisfu) [reply cl/assets/static-global/css/override.css]: Yeah, bootstrap has classes for text with different statuses: `text-success`, `text-danger`, and `text-warning`
Now using them, I added a new template tag filter to easily assign them.

File: cl/users/templates/includes/webhook-event-detail.html
Comment (albertisfu) [reply cl/users/templates/includes/webhook-event-detail.html]: Yes, I moved all this common code to webhooks_base.html so it's not repeated.

File: cl/users/templates/includes/webhooks_htmx/webhooks-pagination.html
Comment (albertisfu) [reply cl/users/templates/includes/webhooks_htmx/webhooks-pagination.html]: Yeah, now using the global pagination after merging the webhook pagination variants:
Screen Shot 2022-11-21 at 11 15 24

File: cl/users/templates/profile/webhooks_logs.html
Comment (albertisfu) [reply cl/users/templates/profile/webhooks_logs.html]: Yes, I moved all this common code to webhooks_base.html so it's not repeated.

File: cl/users/views.py
Comment (albertisfu) [reply cl/users/views.py]: Merged!

**Comment by albertisfu** [comment]
Thanks for the feedback!

These changes are already applied:

1. When we don't have a status code or Next attempt date, we put a dash `-` instead of saying `None`
2. Changed `Date created` to `Event Date` in the webhook logs list and webhook event detail page.
3. On the detail page changed `Event ID`to `Idempotency-key`.
4. About making every log row clickable, due to the <a></a> breaking the table structure we just keep the Detail button, as we talked about.
5. On the detail page, added the python status code label. 
6. As we talked about logs for past events, we just store the data for the last retry, updating the event accordingly. 
Changed the "Try counter" field to "Failure count" on the detail page. The number keeps the same according to the logic set: 
  If a Webhook event is delivered successfully on the first try the `Failure count` would be: 0
  If it fails the failure count would be 1 and enqueued for its retry.  The max Failure count for a webhook event is 8, with 7 possible schedule retries. If the webhook fails after the 7th retry it would be failure number 8 and the webhook will be disabled, with a total elapsed time of 54:39 since the first try.

    Retry Count | New Delay | Elapsed
    -- | -- | --
    1 | 0:03 | 0:03
    2 | 0:09 | 0:12
    3 | 0:27 | 0:39
    4 | 1:21 | 2:00
    5 | 4:03 | 6:03
    6 | 12:09 | 18:12
    7 | 36:27 | 54:39
7. Header on the Webhook detail page was capitalized and pluralized: `Webhook Event Details`
8. About the global Webhook event failure count, as we talked about it, it will keep going up and up to keep track of webhooks with a big failure count after they're re-enabled.

Some related screenshots:

Screen Shot 2022-11-21 at 12 00 41
Screen Shot 2022-11-21 at 12 01 20

9. Changed the error message to: `We ran into an error processing your request and will be looking into it. Please try again later.`
10. Webhook Test modal updated as requested:

Screen Shot 2022-11-21 at 10 53 49
Screen Shot 2022-11-21 at 10 53 54

Let me know if there is anything else here.

File: cl/assets/static-global/js/webhooks-page.js
Comment (mlissner) [inline cl/assets/static-global/js/webhooks-page.js:40]: This still needed?

File: cl/assets/static-global/js/webhooks-page.js
Comment (mlissner) [inline cl/assets/static-global/js/webhooks-page.js:76]: I can't explain why this is better English, but I guess when it's in a title you skip words:

```suggestion
      webhook_title.innerHTML = 'Add webhook endpoint';
```

File: cl/users/templates/includes/webhook-event-detail.html
Comment (mlissner) [inline cl/users/templates/includes/webhook-event-detail.html:10]:
```suggestion
    <h3 class="v-offset-below-2">Webhook Event Details{% if webhook_event.debug %} (Testing){% endif %}</h3>
```

**Review by mlissner — APPROVED** [review]
A couple little things, but I think this is ready. Feel free to merge after tweaks, thank you!

File: cl/assets/static-global/js/webhooks-page.js
Comment (albertisfu) [reply cl/assets/static-global/js/webhooks-page.js]: Yes, now we use it to show the checkbox icon inside the send button once the request is sent. I only updated the comment.

File: cl/assets/static-global/js/webhooks-page.js
Comment (albertisfu) [reply cl/assets/static-global/js/webhooks-page.js]: Changed.

File: cl/users/templates/includes/webhook-event-detail.html
Comment (albertisfu) [reply cl/users/templates/includes/webhook-event-detail.html]: Changed.

**Comment by albertisfu** [comment]
Thanks, these changes were applied, merging...

**Pull request merged**
